### ISSUE TITLE
Failure to compile native grpc library

### ISSUE THREAD

**Issue by hlb8122** [quasarframework/quasar#5605]
**Describe the bug**
The grpc library gives the following error at compile time:
```
WARNING  Compiled with 3 warnings                                                                                                                                 12:45:49 AM

 warning  in ./node_modules/grpc/src/grpc_extension.js

Critical dependency: the request of a dependency is an expression

 warning  in ./node_modules/grpc/node_modules/node-pre-gyp/lib/pre-binding.js

Critical dependency: the request of a dependency is an expression

 warning  in ./node_modules/grpc/node_modules/node-pre-gyp/lib/util/versioning.js

Critical dependency: the request of a dependency is an expression
```

Once the app has booted the following error is raised:
```
[Vue warn]: Error in mounted hook: "Error: Cannot find module '/home/harry/projects/minimal/node_modules/grpc/package.json'"

found in

---> <App> at src/App.vue
       <Root>
```

```
Error: Cannot find module '/home/harry/projects/minimal/node_modules/grpc/package.json'
    at webpackEmptyContext (eval at ./node_modules/grpc/node_modules/node-pre-gyp/lib sync recursive (app.js:3003), <anonymous>:2:10)
    at Object.exports.find (webpack-internal:///./node_modules/grpc/node_modules/node-pre-gyp/lib/pre-binding.js:20)
    at Object.eval (webpack-internal:///./node_modules/grpc/src/grpc_extension.js:29)
    at eval (webpack-internal:///./node_modules/grpc/src/grpc_extension.js:64)
    at Object../node_modules/grpc/src/grpc_extension.js (app.js:3467)
    at __webpack_require__ (app.js:770)
    at fn (app.js:130)
    at eval (webpack-internal:///./node_modules/grpc/src/client_interceptors.js:144)
    at Object../node_modules/grpc/src/client_interceptors.js (app.js:3420)
    at __webpack_require__ (app.js:770)
```

https://github.com/hlb8122/minimal-failure-example

**To Reproduce**
1. Initialize a new project
2. npm i grpc
3. Import and use grpc

**Expected behavior**
Ability to use the library

**Platform:**
OS: Ubuntu 19.04
Node: 10, 11, 12, 13
NPM: 6
Yarn: 1
Browsers: Firefox
Electron: 5

**Additional context**
Tried various permutations of the gRPC library all with no luck. The minimal example linked has only gRPC imported.

**Comment by pdanpdan** [comment]
Why are you logging the issue also here :) ?
We cannot fix gRPC. If it is not compiling then Quasar has nothing to use.

**Comment by hlb8122** [comment]
> Why are you logging the issue also here :) ?
> We cannot fix gRPC. If it is not compiling then Quasar has nothing to use.

https://github.com/grpc/grpc-node/issues/1185#issuecomment-554470465

It seems to be a certain feature lacking in quasar framework. Do you imagine this could be implemented so that it could work?

**Comment by hlb8122** [comment]
Perhaps this is related?
https://github.com/grpc/grpc/issues/11435#issuecomment-377711587

**Comment by pdanpdan** [comment]
Can you try to add the gRPC path to externals using https://quasar.dev/app-extensions/development-guide/index-api#api.chainWebpack?

**Comment by hlb8122** [comment]
Adding
```
        cfg.externals = [
          'grpc'
        ]
```
removes the compile time warning but the browser console error is now

```
[Vue warn]: Error in mounted hook: "ReferenceError: grpc is not defined"

found in

---> <App> at src/App.vue
       <Root> vue.runtime.esm.js:619
```

**Comment by mboeni** [comment]
I have come across a similar issue. Using grpc in a Quasar (electron) project:

`
external "grpc"?a322:1 Uncaught ReferenceError: grpc is not defined
    at eval (external "grpc"?a322:1)
    at Object.grpc (app.js:1238)
    at __webpack_require__ (app.js:791)
    at fn (app.js:151)
    at eval (GatewayService_grpc_pb.js?4c32:4)
    at Module../src/nxs/gwmessages/GatewayService_grpc_pb.js (0.js:11)
    at __webpack_require__ (app.js:791)
    at fn (app.js:151)
    at Module.grpcHelloWorld (utility.js?bcca:19)
    at VueComponent.keyListener (MainLayout.vue?713b:371)
`

Any ideas on this?

**Comment by rstoenescu** [comment]
The initial repo link is not valid anymore (https://github.com/hlb8122/minimal-failure-example). Can you put it back up pls?

Also, Quasar is not lacking anything. The error comes from Electron itself. Electron warns against using an expression as the request of a dependency.

**Comment by hlb8122** [comment]
> The initial repo link is not valid anymore (https://github.com/hlb8122/minimal-failure-example). Can you put it back up pls?

Sorry, just had a clean out and forgot this was open still. It's back up now.

Quickest way to reproduce is:
```bash
quasar create minimal-failure-example
cd minimal-failure-example/
yarn add grpc
```
Replace `App.vue` with
```javascript
<template>
  <div id="q-app">
    <router-view />
  </div>
</template>

<script>
var grpc = require('grpc')

export default {
  created () {
    grpc.getClientChannel() // This can be anything involving gRPC
  }
}
</script>
```
then
```bash
quasar dev
```

> Also, Quasar is not lacking anything. The error comes from Electron itself. Electron warns against using an expression as the request of a dependency.

In this case the error occurs on any platform I attempt.

**Comment by mboeni** [comment]
Same here... -.- The error is quite persistent, no matter what I try...

**Comment by wenoptics** [comment]
> Adding
> 
> ```
>         cfg.externals = [
>           'grpc'
>         ]
> ```
> 
> to `extendWebpack` or adding
> 
> ```
>       chainWebpack: config => {
>         config.externals(
>           {
>             'grpc': 'grpc'
>           }
>         )
>       },
> ```
> 
> removes the compile time warning but the browser console error is now
> 
> ```
> [Vue warn]: Error in mounted hook: "ReferenceError: grpc is not defined"
> 
> found in
> 
> ---> <App> at src/App.vue
>        <Root> vue.runtime.esm.js:619
> ```

@hlb8122 @mboeni I think this issue may be related to running gRPC - which is a native module - on the renderrer process. Try `require('grpc')` in main process (e.g. `electron-main.js`) should work fine.

Tell webpack to exclude gRPC by 

```javascript
cfg.externals = [
      'grpc'
]
```

and try:

```javascript
// In electron-main.js
console.log(require('grpc'))
```

should work.

---

If you still want to use grpc in renderer process, I guess you can still inject the grpc object before the redered process with [`preload` provided by Quasar](https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script).

**Comment by rstoenescu** [comment]
@wenoptics is right

**Comment by asymness** [comment]
I was able to use gRPC with a simple Electron-vue project (https://github.com/SimulatedGREG/electron-vue) in the renderer process without any workarounds. However, when I try to do the same in a Quasar-generated application, it gives me `module not found` error. After adding `grpc` to `cfg.externals`, I get `grpc is not defined.`

Is there any workaround to get `grpc-node` working with Quasar in renderer process. Was anyone able to get this working?

**Comment by wenoptics** [comment]
@asymness What I did is inject the grpc object via the [preload script](https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script#Limitations-on-electron-preload.js). Say 

```javascript
window.grpc = grpc
```

then it should work in the renderer process by accessing `window.grpc`. 

Maybe worthy to mention that this indeed is a workaround/hack (you will also find Warnings from Electron), no feeling that this is a proper implementation (but works for me).

**Comment by asymness** [comment]
Hi, thanks for your response. I did try this workaround as mentioned by you earlier, but I was getting `exports is not defined` error. Maybe it had to do something with Webpack, and I'm not too familiar with it.

I was able to solve this problem and get this working by creating a brand-new Vue CLI project using `vue create`, then added Quasar `vue add quasar`, added electron-builder `vue add electron-builder`, upgraded electron version and then installed `grpc`. Using this method, I was able to get the grpc-node module working with renderer process of electron without any workaround.

I still have no idea why this wasn't working with a Quasar CLI generated app...
